### ISSUE TITLE
Error while setup: "Unable to connect, please check serial port and electrical connection"

### ISSUE THREAD

**Issue by sebi-alt** [NECH2004/smartmeter_austria#2]
## System Information

version | core-2023.5.4
-- | --
installation_type | Home Assistant OS
dev | false
hassio | true
docker | true
user | root
virtualenv | false
python_version | 3.10.11
os_name | Linux
os_version | 6.1.21-v8
arch | aarch64
timezone | Europe/[redacted]
config_dir | /config

<details><summary>Home Assistant Cloud</summary>

logged_in | false
-- | --
can_reach_cert_server | ok
can_reach_cloud_auth | ok
can_reach_cloud | ok

</details>

<details><summary>Home Assistant Supervisor</summary>

host_os | Home Assistant OS 10.1
-- | --
update_channel | stable
supervisor_version | supervisor-2023.06.2
agent_version | 1.5.1
docker_version | 23.0.3
disk_total | 219.4 GB
disk_used | 10.3 GB
healthy | true
supported | true
board | rpi4-64
supervisor_api | ok
version_api | ok
installed_addons | ESPHome (dev) (dev), Samba share (10.0.0), Terminal & SSH (9.7.0), Studio Code Server (5.5.6), FTP (4.7.0)

</details>

<details><summary>Dashboards</summary>

dashboards | 1
-- | --
resources | 0
views | 4
mode | storage

</details>

<details><summary>Recorder</summary>

oldest_recorder_run | May 29, 2023 at 12:02
-- | --
current_recorder_run | June 18, 2023 at 16:41
estimated_db_size | 84.33 MiB
database_engine | sqlite
database_version | 3.40.1

</details>

### Checklist

- [ ] I have enabled debug logging for my installation.
- [ ] I have filled out the issue template to the best of my ability.
- [ ] This issue only contains 1 issue (if you have multiple issues, open one issue for each issue).
- [ ] This issue is not a duplicate issue of currently [previous issues](https://github.com/ludeeus/integration_blueprint/issues?q=is%3Aissue+label%3A%22Bug%22+)..

### Describe the issue

Thank you for providing this integration. I am running into some issues during setup and would be very thankful for any help.

During setup, the integration UI fails with the prompt "Unable to connect, please check serial port and electrical connection". According to [smartmeter_austria_energy/src/smartmeter_austria_energy/exceptions.py](https://github.com/NECH2004/smartmeter_austria_energy/blob/dev/src/smartmeter_austria_energy/exceptions.py) this is due to an unexpected byte stream.

What I did for troubleshooting:
- Switch the wires going from smartmeter to the MBUS to USB dongle
- Check the connection of these wires
- Check if the USB device appears in Hardware info of homeassistant at /dev/ttyUSB0
- Reboot homeassistant
- Recheck the encryption key provided by energy suppliers website

I am using [this](https://www.amazon.de/dp/B0BXPRCWK9) MBUS to USB converter. Unfortunately, the images of the amazon page did not match with the delivered product, so [there](https://www.amazon.de/dp/B0BSKNGN51/) are some images of the actual device.

I installed the addon by copying the contents of custom_components (not via hacs)

As far as I am concerned, I got a bad USB dongle. Is there anything I missed? Thanks for any help!

### Reproduction steps

1. Install the integration by adding copying all contents of custom_components to your custom_components directory
2. Enable the customer interface of your smartmeter and get your encryption key via the website of your energy supplier
3. Purchase [this](https://www.amazon.de/dp/B0BXPRCWK9) USB to MBUS dongle
4. Add the integration

### Debug logs

```text
This error originated from a custom integration.

Logger: custom_components.smartmeter_austria.config_flow
Source: custom_components/smartmeter_austria/config_flow.py:52
Integration: Smart Meter Austria (documentation, issues)

Could not connect to device=/dev/ttyUSB0
```

**Comment by NECH2004** [comment]
Hello sebi-alt,
it seems that Homeassistant is able to find your dongle. If you look into your hardware configuration (http://homeassistant.local:8123/config/hardware) you should be able to find your dongle. It should be your Prolific converter.
Normally the error occurs if opening the data stream is possible but decoding the data is not possible.
This could happen if your key is wrong and decoding returns garbage or your stream is not as supposed.
Mind that copy/paste sometimes changes characters (eg. quotation marks).

Reading from the adapter can only be done from one instance, perhaps an other program uses the device?. Can HA really read from the device?. I'm running HA-OS in a Proxmox environment, the VM runs privileged.

Also possible that your provider has changed its protocol.
Which provider/smart meter do you use?

A simple way to find your problems would be to test your USB-device on your desktop PC.
My implementation repo (https://github.com/NECH2004/smartmeter_austria_energy) that mainly is based on Stefan's code would be a good starting point. It contains all functions that are used by HA.

Just setup a Python development environment and  connect your M-Bus dongle into your (windows) PC.
The sample should show how to connect, debugging the Smartmeter class and the read() method should be quite easy.

Hope this helps.

**Comment by sebi-alt** [comment]
Hello NECH2004,
thank you for this fast response and sorry for not responding for a while. 

The key is definitely correct and homeassistant labels the dongle as an `usb-Prolific_Technology_Inc._USB-Serial_Controller_CPEGb118D16-if00-port0`. My only instance is running on a Raspberry 4 using HA-OS. The smartmeter is of type `Honeywell DM515`. 

On my Windows PC the basic script in the readme of [the implementation repo](https://github.com/NECH2004/smartmeter_austria_energy/blob/2f7c53f53468f2eed7c18f8cca44cad64435252e/README.md?plain=1#L35) raises an error in smartmeter.py: [`SmartmeterTimeoutException` in line 138](https://github.com/NECH2004/smartmeter_austria_energy/blob/2f7c53f53468f2eed7c18f8cca44cad64435252e/src/smartmeter_austria_energy/smartmeter.py#L138C1-L138C55) caused by [`SmartmeterException` in  line 148](https://github.com/NECH2004/smartmeter_austria_energy/blob/2f7c53f53468f2eed7c18f8cca44cad64435252e/src/smartmeter_austria_energy/smartmeter.py#L147). What could be the reason for this timeout issue?

Any ideas on how to proceed? Again, thank you for your fast and excellent response

**Comment by NECH2004** [comment]
Hello Sebi,
sorry for the delay (vacation)...

Here is a python test file doing some calls against the serial MBUS adapter and helps to debug your problem.
The test file _sometimes_ fails executing test step no. 6 if executed on one of my Windows PC leading to the same error.
It seems that the default wait interval of 1 s could be critical.
Perhaps the reading intervals of the Smart Meters are inaccurate or the timing between the Smart Meter types is different..
If step 1 to 5 are OK, you can try to add the additional parameter 'interval' to the SmartMeter constructor.
A value of 2 (max 5) seconds fixes (my) timeout problem.
Hope it helps.


# 1. plug in usb adapter in to PC: Power LED lights up
# 2. plug in M-Bus cable into Smart Meter (KAIFA 309M): RX LED blinks (about 2x every 5 s)


# 3. Find your COM Port:
#    All found COM ports are listed in brackets.
import serial.tools.list_ports

com_ports = serial.tools.list_ports.comports(include_links=True)
com_ports_list = []
for port in com_ports:
    com_ports_list.append(port.device)

if len(com_ports_list) > 0:
    print(com_ports_list)

# use first com port for test:
used_com_port = com_ports_list[0]

#4. test device

#SERIAL_PORT = '/dev/ttyAMA0'
#SERIAL_PORT = 'COM3'

SERIAL_PORT = used_com_port
print(f"used COM port: {used_com_port}")

# be sure to set this to the same rate used on the Smart Meter
SERIAL_RATE = 2400
ser = serial.Serial(SERIAL_PORT, SERIAL_RATE)

# using ser.readline() assumes each line contains a single reading
reading = ser.readline() #.decode('utf-8') # first line
reading2 = ser.readline() #.decode('utf-8') # second line, if reading starts during sending.
ser.close();

# reading is a string...do whatever you want from here
#print(reading)
print(' '.join('{:02x}'.format(x) for x in reading))
print(' '.join('{:02x}'.format(x) for x in reading2))

# look into your output:
# try to find your start bytes of the first data frame in the output:
#   68 fa fa 68
# example:
# ... 7e 2b 56 44 96 27 64 16 _68 fa fa 68_ 53 ff 00 01 67 db 08 4b 46 4d 10 20 03 e1 27 82 01 55 ...

# try to find your start bytes of the second data frame in the output:
#   68 72 72 68 (TINETZ / SALZBURGNETZ) or
#   68 14 14 68 (EVN)


# 6. Try out implementation that is used in Home Assistant:
# mostly the same script as in the readme.

from smartmeter_austria_energy.smartmeter import Smartmeter
from smartmeter_austria_energy.supplier import SUPPLIER_SALZBURGNETZ_NAME, SUPPLIER_EVN_NAME, SUPPLIER_TINETZ_NAME

# your provider
supplier_name = SUPPLIER_SALZBURGNETZ_NAME

# your key
key_hex_string = "-- this is your key --"
port = used_com_port

smartmeter = Smartmeter(supplier_name, port, key_hex_string)

# additional interval param:
#smartmeter = Smartmeter(supplier_name, port, key_hex_string, interval=2)

my_obisdata = smartmeter.read()

print(f"RealEnergyIn: {my_obisdata.RealEnergyIn.ValueString}")
print(f"RealEnergyOut: {my_obisdata.RealEnergyOut.ValueString}")

**Comment by sebi-alt** [comment]
I hope you had a great vacation and thank you for coming back to me with such excellent help
After finding all significant start bytes with the test file I was sure that my hardware was working as expected. The read operation took up to 10 seconds so the `TimeOutException` was justified. After removing [this else-statement](https://github.com/NECH2004/smartmeter_austria_energy/blob/2f7c53f53468f2eed7c18f8cca44cad64435252e/src/smartmeter_austria_energy/smartmeter.py#L137) I was able to get the data from my smartmeter. Raising the interval to 2 did not work, but since the interval influenced the if-statement, I decided to try it with `interval=4` which solved it.

I locally changed it in the homeassistant component [here](https://github.com/NECH2004/smartmeter_austria/blob/4c383ae5ec52c9ba62cd1bc97d2b64f34e4d4783/custom_components/smartmeter_austria/__init__.py#L51C17-L51C19). The integration works smoothly now. 

Thank you very much for your help
